### PR TITLE
Translation of QGIS project content in Lizmap Web Client

### PR DESCRIPTION
**Pull request by RobiFag**
**Translation**
In this folder there are:

1. a python script to retrieve translatable strings from a QGIS project and its related .cfg file created with the lizmap plugin
2. a javascript file which translates the project content in Lizmap-Web-Client according to the browser language

**The python script (get_translatable_string.py)**
This python script retrieves all possible translatable strings (layers name, groups name, layouts name, fields or aliases name, etc.) from the QGIS project and the related .cfg file. It saves a json file in which the key is the retrieved string and the value should contain the translation of the string. The json file is automatically saved in the media folder. Following an example of the json structure:

`{"Original layer name": "", "Original group name": "", ..}`

The translation has to be provided manually by editing the file and putting the translated string in the double-quotes. Following an example of the translated json:

`{"Original layer name": "Translated layer name", "Original group name": "Translated group name", ..}`

**The js script (translation.js)**
This javascript script translates layers and groups names, the title of the project, the print layouts names and the aliases or fileds names according to the language of the browser and using the json file created with the python script get_translatable_string.py If a translated string is provided in the json file the original string is translated, otherwise the original string is shown in lizmap web client. The json file must be saved in the media folder which needs to be available from the web. Therefore it is necessary to create a symbolic link on the apache directory (e.g. /var/www/html/) to the media folder in the user repository. The javascript retrieves the original string (key) and the translated string (value) from the json file. Then for each key found in the jason check the text of the provided html selector (e.g. `$("div#header div#title h1")`). If the two strings (the one from the json file e the one found in the html page) match the text of the provided html selector is replaced with the translated string, if provided in the json file.

### CODE REVIEW

**Comment by rldhont** [comment]
@mdouchin @nboisteault can you review the code ?

**Comment by rldhont** [comment]
@RobiFag can you add some descritpion to your PR ?

**Comment by RobiFag** [comment]
@rldhont description added!

File: translation/translation.js [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+    uicreated: function(e) {
+        // get browser language
+        var userLang = navigator.language || navigator.userLanguage;
+        if (userLang === 'fr' || userLang === 'fr-FR'){      
+            
+            var translationIT_FR = {};
+
+            // get the json file with transalted strings
+            // the path to the json file must be provided as relative path, using the web url the browser can return a security error
+            fetch('../../../../../../concerteaux3d/translation.json').then(function(response) {
```
Comment (nboisteault) [inline translation/translation.js]: Could you put the URL in a variable at top of the script ?

File: translation/translation.js [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+    lizmapeditionformdisplayed: function(e) {
+        // get browser language
+        var userLang = navigator.language || navigator.userLanguage;
+        if (userLang === 'fr' || userLang === 'fr-FR'){
+
+            var translationIT_FR2 = {};
+
+            // get the json file with transalted strings
+            // the path to the json file must be provided as relative path, using the web url the browser can return a security error
+            fetch('../../../../../../concerteaux3d/translation.json').then(function(response) {
```
Comment (nboisteault) [inline translation/translation.js]: You could cache translation.js as you have download it with "uicreated" event above.

File: translation/translation.js [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+    lizmappopupdisplayed: function(e) {
+        // get browser language
+        var userLang = navigator.language || navigator.userLanguage;
+        if (userLang === 'fr' || userLang === 'fr-FR'){
+
+            var translationIT_FR3 = {};
+
+            // get the json file with transalted strings
+            // the path to the json file must be provided as relative path, using the web url the browser can return a security error
+            fetch('../../../../../../concerteaux3d/translation.json').then(function(response) {
```
Comment (nboisteault) [inline translation/translation.js]: You could cache translation.js as you have download it with "uicreated" event above.

File: translation/translation.js [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+lizMap.events.on({
+    lizmapswitcheritemselected: function(){
+        var userLang = navigator.language || navigator.userLanguage;
+        if (userLang === 'fr' || userLang === 'fr-FR'){        
+            
+            var translationIT_FR = {};
+
+            // get the json file with transalted strings
+            // the path to the json file must be provided as relative path, using the web url the browser can return a security error
+            fetch('../../../../../../concerteaux3d/translation.json').then(function(response) {
```
Comment (nboisteault) [inline translation/translation.js]: You could cache translation.js as you have download it with "uicreated" event above.

File: translation/translation.js [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+                    Object.keys(translationIT_FR3).forEach((italian) => {
+                        //check if a translation is provided for each string
+                        if (translationIT_FR3[italian] != ""){
+                            // translate layers name in the popup
+                            $("div#content div#map-content div#map.olMap div#OpenLayers_Map_377_OpenLayers_ViewPort div#OpenLayers_Map_377_OpenLayers_Container div#liz_layer_popup.olPopup.lizmapPopup div#liz_layer_popup_GroupDiv div#liz_layer_popup_contentDiv.olPopupContent.lizmapPopupContent h4").filter(function(){
+                                return $(this).text() === italian
+                            }).html(translationIT_FR3[italian]);
+                            // translate aliases or fields name in the popup
+                            $("div#content div#map-content div#map.olMap div#OpenLayers_Map_377_OpenLayers_ViewPort div#OpenLayers_Map_377_OpenLayers_Container div#liz_layer_popup.olPopup.lizmapPopup div#liz_layer_popup_GroupDiv div#liz_layer_popup_contentDiv.olPopupContent.lizmapPopupContent div.lizmapPopupDiv table.lizmapPopupTable tbody tr th").filter(function(){
```
Comment (nboisteault) [inline translation/translation.js]: Are you sure 377 in `#OpenLayers_Map_377_OpenLayers_Container` won't change when reloading ?

File: translation/translation.js (hunk 1 above)
Comment (mdouchin) [reply translation/translation.js]: We should use Lizmap Web Client media capabilities here I think. If lizmap is accessible via
https://demo.lizmap.com/lizmap/index.php
you can fetch for example the file `media/photo/beluga.jpg` build an URL like this one 
https://demo.lizmap.com/lizmap/index.php/view/media/getMedia?repository=demo&project=cats&path=media/photo/beluga.jpg

To build this URL, you can use 

```javascript
var mediaUrl = OpenLayers.Util.urlAppend(
     lizUrls.media,
     OpenLayers.Util.getParameterString({
         "repository": lizUrls.params.repository,
         "project": lizUrls.params.project,
         "path": "media/photo/beluga.jpg"
     })
);
```

File: translation/translation.js [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+a symbolic link on the apache directory (e.g. /var/www/html/) to the media folder in the user repository.*/
+
+//This function is executed when the uicreated event is triggered
+lizMap.events.on({
+    uicreated: function(e) {
+        // get browser language
+        var userLang = navigator.language || navigator.userLanguage;
+        if (userLang === 'fr' || userLang === 'fr-FR'){      
+            
+            var translationIT_FR = {};
```
[the remark this answers is not in the document]
Comment (RobiFag) [reply translation/translation.js]: Sure, no problem! Maybe something like 'translated_string'?

**Comment by Gustry** [comment]
Nice idea.

For curiosity, did you have a look to the translation of a QGIS project ? https://changelog.qgis.org/en/qgis/version/3.4-LTR/#translation-of-qgis-projects Would it be hard to export from TS files to JSON?

File: translation/translation.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (RobiFag) [reply translation/translation.js]: I changed all variables using more generic names.

File: translation/translation.js (hunk 1 above)
Comment (RobiFag) [reply translation/translation.js]: I have implemented the solution suggested by @mdouchin and it works without the symbolic link on the apache directory.

File: translation/translation.js (hunk 5 above)
Comment (RobiFag) [reply translation/translation.js]: The id `#OpenLayers_Map_377_OpenLayers_Container` is always the same in my browser, could it be browser dependent?

**Comment by RobiFag** [comment]
> Nice idea.
> 
> For curiosity, did you have a look to the translation of a QGIS project ? https://changelog.qgis.org/en/qgis/version/3.4-LTR/#translation-of-qgis-projects Would it be hard to export from TS files to JSON?

Yes sure, I have also work on a script to update the .ts file created by QGIS (see https://github.com/qgis/QGIS-Enhancement-Proposals/issues/90 for more details). To my knowledge, it is surely possible to export an xml (like .ts) file to a json in python or javascript (using standard libraries). However, if I understand correctly, it is necessary to check if all the information needed for translation in lizmap web client are provided and it is also necessary to filter out possible useless information.

File: translation/translation.js (hunk 2 above)
Comment (RobiFag) [reply translation/translation.js]: I have changed the code, now the fetch of the translation.json file is done only once in the "uicreated" event.

File: translation/translation.js (hunk 3 above)
Comment (RobiFag) [reply translation/translation.js]: I have changed the code, now the fetch of the translation.json file is done only once in the "uicreated" event.

File: translation/translation.js (hunk 4 above)
Comment (RobiFag) [reply translation/translation.js]: I have changed the code, now the fetch of the translation.json file is done only once in the "uicreated" event.

**Comment by RobiFag** [comment]
I have already updated the python script. Now if the .json file already exists it will be updated only with the new strings.

**Comment by RobiFag** [comment]
Hi @rldhont @nboisteault @mdouchin !
We resolved all your comments and added the update of the json file in order to make the python script more usable.

Is it possible to accept and close the pull request?

File: translation/get_translatable_string.py [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+
+#retrieves the aliases or the fields names of all vector layer in a specified group
+#and put them in the related dictionary.
+#It works over the whole layer tree too. Comment the lines related to the specified group
+#to retrieve the alias/fields name of all vector layers in the project
+root = projectInstance.layerTreeRoot()
+sel_gr = 'ConcertEaux' #the name of the desired group, to be commented if not necessary 
+for child in root.children():
+    if isinstance(child, QgsLayerTreeGroup):
+        if child.name() == sel_gr: #to be commented if not necessary
```
Comment (Gustry) [inline translation/get_translatable_string.py]: To put in a variable at the top of the file?

File: translation/translation.js [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+                    $("div#content div#sub-dock div.sub-metadata div.menu-content dl dd").filter(function(){
+                        return $(this).text() === inputLang
+                    }).html(inputLang);
+                }
+            });
+        }
+    }
+});
+
+// Work in progress: translate the layer tooltip that is shown when the mouse pointer pass over the layer name in the layer tree (see issue https://github.com/3liz/lizmap-web-client/issues/428)
```
[the remark this answers is not in the document]
Comment (RobiFag) [reply translation/translation.js]: Yes it is still WIP, see https://github.com/3liz/lizmap-web-client/issues/428

File: translation/get_translatable_string.py [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+            for gr in child.children():
+                if isinstance(gr, QgsLayerTreeGroup):
+                    lyrs_tmp = gr.findLayers()
+                    for lt in lyrs_tmp:
+                        if isinstance(lt.layer(), QgsVectorLayer):
+                            lyrs.append(lt)
+                elif isinstance(gr.layer(), QgsVectorLayer):
+                    lyrs.append(gr)
+
+#ATTENTION: the attributeAliases() method does not read aliases of layers added during the active session of QGIS.
```
[the remark this answers is not in the document]
Comment (RobiFag) [reply translation/get_translatable_string.py]: I have put this comment in the description of the script too.

**Comment by RobiFag** [comment]
> Thanks for your work.
> Just some suggestions :
> 
> * use markdown https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#code
> * make all configurations set by variables at the top of the file. Users doesn't know much about JS or Python, they won't read the code itself.

Thanks @Gustry for your suggestions..I hope I have correctly understood and hence resolved them in the proper way.

**Comment by Gustry** [comment]
- [ ] Use the squash button
- [ ] Create a ticket on Lizmap documentation ?

**Comment by RobiFag** [comment]
> Thanks, it looks fine for me.
> I can't comment on the Javascript side, I will let my colleagues on that part.
> 
> Side question, which LWC version are you using for this?

We have tested the translation scripts with the versions 3.3.4 and 3.3.5

**Comment by RobiFag** [comment]
> * [ ]  Use the squash button

What do you mean for 'squash button'?

> * [ ]  Create a ticket on Lizmap documentation ?

Sure, but maybe it is better to create an ad hoc section..what do you think?

File: translation/get_translatable_string.py [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+            print('Unable to update the .json file')
+    else:
+        print('The .json file already contains all translatable strings. It will not be updated.')
+else:
+    try:
+        with open(jsonFilePath, 'w', encoding='utf-8') as fp:
+            json.dump(translation, fp, ensure_ascii=False)
+        print('The .json file has been saved')
+    except:
+        print('Unable to save the .json file')
```
Comment (Gustry) [inline translation/get_translatable_string.py]:
```suggestion
        print('Unable to save the .json file to "{}".'.format(jsonFilePath))
        print('Check that the folder exists.')
```

**Comment by Gustry** [comment]
> What do you mean for 'squash button'?

Don't worry, it's a reminder for us. You made a PR with 18 commits, we will squash it on github.

> Sure, but maybe it is better to create an ad hoc section..what do you think?

Well, until this is not in core, this help can stay here I think.

I just tried your script.
The resulting file contains a single line with all translations on it.
This is not recommanded because:
* difficult to read
* If you use a VCS system such as GIT, it makes it impossible to track translations.

In the python `json` module, there is a function to format a JSON string. It will add indentation etc. Can you use it?

I tried on different Lizmap projects and I **always** have a string with `\n` and a lot of spaces, the last one. Is-it normal?
```
{"\n          Cats movements\n        ": ""}
```

File: translation/get_translatable_string.py (hunk 10 above)
Comment (RobiFag) [reply translation/get_translatable_string.py]: Done!

**Comment by RobiFag** [comment]
> > What do you mean for 'squash button'?
> 
> Don't worry, it's a reminder for us. You made a PR with 18 commits, we will squash it on github.
> 
> > Sure, but maybe it is better to create an ad hoc section..what do you think?
> 
> Well, until this is not in core, this help can stay here I think.
> 
> I just tried your script.
> The resulting file contains a single line with all translations on it.
> This is not recommanded because:
> 
> * difficult to read
> * If you use a VCS system such as GIT, it makes it impossible to track translations.
> 
> In the python `json` module, there is a function to format a JSON string. It will add indentation etc. Can you use it?

I added an `indent=4` in the dump function parameters. now the json file showld be better formatted.

> 
> Sorry, I haven't tried the Javascript part. So only file translated file is possible? (single `translation.json` with a single language)
> 

I'm not a javascript expert and right now the only possible way I can suggest at the moment would be duplicating the code. Obviously I'm sure there are better solutions and we are very glad to discuss them with you whenever possible.

> I tried on different Lizmap projects and I **always** have a string with `\n` and a lot of spaces, the last one. Is-it normal?
> 
> ```
> {"\n          Cats movements\n        ": ""}
> ```

Yes it is normal! This is the line to translate the title. If you try to inspect the title on your lismap project, you'll see that the value of the `div#title h1` has \n and spaces. In order to translate the title, it is necessary to pass the whole string including \n and spaces.

**Comment by RobiFag** [comment]
> Sorry, I haven't tried the Javascript part. So only file translated file is possible? (single `translation.json` with a single language)

I all, I have just uploaded a new version of both python and javascript files. Now the python file creates a json for each language specified in a list (e.g translation_en.json,  translation_fr.json, etc.). On the javascript side, the file translation.js loads the corresponding json file depending on the browser language.

**Comment by rmarzocchi84** [comment]
Let us know if it is ok  for you. I think it is a good work which in future can be also integrated in the lizmap-plugin.

**Pull request merged**